freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/RebootInstancesService.java

```diff
@@ -0,0 +1,111 @@
+                LOGGER.info("Unknown instance ID {} in request", id);
+            }
+        };
+        return Stream.ofNullable(request.getInstanceIds()).flatMap(Collection::stream)
+                .peek(logit)
+                .filter(id -> !allInstances.contains(id))
+                .collect(Collectors.toSet());
+    }
+
+    private Map<String, InstanceMetaData> getInstancesToReboot(Map<String, InstanceMetaData> allInstances,
+            List<String> instanceIds, Map<String, InstanceStatus> health) {
+        Consumer<Map.Entry> logit = entry -> {
+            if (health.get(entry.getKey()) != null && health.get(entry.getKey()).isAvailable()) {
+                LOGGER.info("Not rebooting AVAILABLE instance {} because force reboot was not selected.", entry.getKey());
+            }
+        };
+        return allInstances.entrySet().stream()
+                .peek(logit)
+                .filter(entry -> instanceIds == null || instanceIds.isEmpty() || instanceIds.contains(entry.getKey()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private Map<String, InstanceMetaData> getAllInstancesMap(Stack stack) {
+        return stack.getInstanceGroups().stream().flatMap(instanceGroup -> instanceGroup.getInstanceMetaData().stream())
+                .collect(Collectors.toMap(InstanceMetaData::getInstanceId, Function.identity()));
+    }
+
+    public void rebootInstances(String accountId, RebootInstancesRequest request) {
+        // If no instance passed in request, reboot all bad instances
+        // If instances passed in request, reboot all valid passed bad instances
+        // If force and instances passed in request, reboot all valid passed instances
+        // If force and no instances passed in request, reboot all instances
+        Stack stack = stackService.getByEnvironmentCrnAndAccountIdWithLists(request.getEnvironmentCrn(), accountId);
+        MDCBuilder.buildMdcContext(stack);
+        Map<String, InstanceMetaData> allInstances = getAllInstancesMap(stack);
+        Set<String> invalidInstanceIds = getInvalidInstances(allInstances.keySet(), request);
+        if (!invalidInstanceIds.isEmpty()) {
+            String msg = MessageFormat.format("Invalid instanceIds are present in request: {0}.", String.join(", ", invalidInstanceIds));
+            LOGGER.error(msg);
+            throw new BadRequestException(msg);
+        }
+
+        Map<String, InstanceStatus> health = getInstanceHealth(accountId, request);
+
+        Map<String, InstanceMetaData> instancesToReboot = getInstancesToReboot(allInstances, request.getInstanceIds(), health);
+        for (Map.Entry<String, InstanceMetaData> instance: instancesToReboot.entrySet()) {
+            instanceMetaDataService.updateStatus(stack, instance.getKey(), InstanceStatus.REBOOTING);
+            flowManager.notify(REBOOT_EVENT.event(), new InstanceEvent(REBOOT_EVENT.event(), instance.getValue().getId()));
+        }
```

Comment:
why do we start a new flow for every instance? I think we should start 1 flow with all the instance IDs which we would like to restart.

Reply:
We eventually want the ability to do a rolling reboot of the servers instead of rebooting all at the same time.  My understanding is the way to accomplish this is by implementing a custom FlowEventChainFactory that can dynamically tie the events together so we'd need separate events to achieve that.

Reply:
I think we should have the flow capable of restarting multiple instance at a time. If we have 2 unhealthy node from 5 we should restart them together. If we would like to restart all, we can still do it in 2 batch. 
If all node is in a bad state does it have any benefit to reboot them in a rolling mode?

Reply:
This had been refactored.  However, we will be adding rolling reboot later.  One of the uniqueness here is that we're running multiple services per instance.  If one service, say DNS, is down on one node, and PKI is down on another, both would be considered bad, but we'd want to rolling reboot these so we don't have total loss of service.